### PR TITLE
Trim System.Title if length is > 128

### PR DESCRIPTION
**Pull request by kananb**
## Summary of the Pull Request
Trim System.Title when longer than 128 characters. Trims some extra characters to append the first 8 chars of the title hash for collision avoidance.

We probably want to write the full System.Title value somewhere in the bug so that the trimmed information isn't just lost.

_What is this about?_
https://github.com/microsoft/onefuzz/issues/3015

## PR Checklist
* [ ] Applies to work item: #xxx
* [ ] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/onefuzz) and sign the CLI.
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

_What does this include?_

## Validation Steps Performed

_How does someone test & validate?_

### CODE REVIEW

File: src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs [hunk 1]
```diff
@@ -396,6 +400,15 @@ public sealed class AdoConnector {
 
                 if (string.Equals(field, "System.Tags")) {
                     value += ";Onefuzz";
+                } else if (string.Equals(field, "System.Title") && value.Length > MAX_SYSTEM_TITLE_LENGTH) {
+                    var titleHashString = System.Security.Cryptography.SHA256.HashData(Encoding.UTF8.GetBytes(value))
+                        .Aggregate(
```
Comment (Porges) [inline src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs:405]: You can use `Convert.ToHexString` here

**Comment by Porges** [comment]
Agree about this:
> We probably want to write the full System.Title value somewhere in the bug so that the trimmed information isn't just lost.

Could we add an additional field to the template for the full title and always include it?

File: src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs (hunk 1 above)
Comment (kananb) [reply src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs:405]: To replace the hash? Wouldn't that risk collision for titles starting with the same first 128 chars?

**Comment by kananb** [comment]
> Could we add an additional field to the template for the full title and always include it?
Something like a unique field with the full title in it? Where are these and the other non-custom fields visible?

**Review by chkeita — APPROVED** [review]

**Comment by Porges** [comment]
> > Could we add an additional field to the template for the full title and always include it?
> 
> We can probably put it at the top of the bug's "Repro Steps". I don't think we have much control over the fields since they're org specific.

Sorry, I was imprecise here, I meant a variable for the Scriban template to render, not the ADO fields.

File: src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs (hunk 1 above)
Comment (Porges) [reply src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs]: To replace the conversion to hex:

```cs
hash.Aggregate(
  new StringBuilder(),
  (builder, next) => builder.Append(next.ToString("x2")),
  builder => builder.ToString())
```
  
Is equivalent to (aside from being in a different case):
```cs
Convert.ToHexString(hash)
```

**Review by Porges — APPROVED** [review]

File: src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs (hunk 1 above)
Comment (kananb) [reply src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs]: Thanks for the clarification! I probably should have been able to figure that `ToHexString` would be for the hex string conversion part.

**Comment by mgreisen** [comment]
@Kananb how did we decide on the 128 char limit?

**Comment by kananb** [comment]
> @kananb how did we decide on the 128 char limit?

https://github.com/MicrosoftDocs/azure-devops-docs/issues/5890#issuecomment-539632059

File: src/ApiService/ApiService/onefuzzlib/notifications/NotificationsBase.cs [hunk 2]
```diff
@@ -85,6 +87,7 @@ public class Renderer {
             return new Renderer(
                 container,
                 filename,
+                issueTitle,
```
Comment (tevoinea) [inline src/ApiService/ApiService/onefuzzlib/notifications/NotificationsBase.cs:90]: Let's render the title, untrimmed.

**Pull request merged**